### ISSUE TITLE
Proposal: Remove Maui's Dependency on Extensions.Hosting

### ISSUE THREAD

**Issue by eerhardt** [dotnet/maui#4393]
### Description

The default startup logic in a Maui app uses `MauiAppBuilder` to register DI services, including configuration sources and logging providers. This design follows a similar programming model in ASP.NET apps - both server side and Blazor WASM apps. The default "dotnet new maui" template generates `MauiProgram.cs` file that looks like:

```C#
public static class MauiProgram
{
    public static MauiApp CreateMauiApp()
    {
        var builder = MauiApp.CreateBuilder();
        builder
            .UseMauiApp<App>()
            .ConfigureFonts(fonts =>
            {
                fonts.AddFont("OpenSans-Regular.ttf", "OpenSansRegular");
            });

        return builder.Build();
    }
}
```

This makes sense and is familiar to ASP.NET developers.

However, one public API hanging off of the `MauiAppBuilder` class is a `public IHostBuilder Host` property. This comes from the `Microsoft.Extensions.Hosting.Abstractions` library. `Microsoft.Extensions.Hosting` is an "app model" itself. It defines a certain order in which a "Hosted Application" is built. For example:

1. There is a "Host Configuration" step, which can be used to configure the host itself. Or it can be used to read "phase-one" configuration that can be used to initialize more configuration in a later step - "App Configuration" (phase-two).
2. Initialization of the `HostEnvironment` and `HostBuilderContext`.
3. An "App Configuration" step which you can add your application configuration sources to.
4. A step to configure any DI services in the Hosted Application.
5. Finally, a way to replace which DI container implementation is used - by default it uses Microsoft.Extensions.DependencyInjection, but apps can opt into using Unity, Autofac, LightInject, etc.

A HostBuilder builds a "Host", which can then be "Started" and "Stopped". Starting and stopping a Host will start and stop any registered "Hosted Services".

Digging into what capabilities the `Microsoft.Extensions.Hosting` provides, I do not believe it is a good fit for mobile / client applications. Even a default ASP.NET Blazor Web Assembly application doesn't use `Microsoft.Extensions.Hosting`. I assume it is the for similar reasons I list below.

1. Extensions.Hosting is really meant for long running processes (like a Windows service, web server, background process, etc). It has a "Lifetime" concept, but it is really simplistic - it has Started, Stopping, and Stopped. This can't model the [App lifecycle](https://docs.microsoft.com/en-us/dotnet/maui/fundamentals/app-lifecycle) needed for a client application.

2. Extensions.Hosting has an "Environment" concept, which are things like "Development", "Staging", and "Production". And while this concept could map to client apps, the way these environments change in Hosted Apps is through Configuration - ex. environment variable, command line parameter, appsettings.json file, etc. My understanding is that a Mobile app doesn't get a chance to change these kinds of configurations once it is built. How do you change a .json file in your already built `.apk` file?

3. Extensions.Hosting has a "ContentRoot", which for a web application points to the folder where it serves files from - like `index.html`, etc. Maui applications don't have "Content" files that get served / shown by the application. Instead the Views are embedded in the app itself. So this concept is useless in a default Maui app.

4. Extensions.Hosting's main concept is its "Hosted Services" or BackgroundService. While this is definitely a useful concept in a client application, the abstraction isn't correct for client apps. Hosted Services get "Start" and "Stop". But what if the app is getting deactivated? Can the service tell the difference between that and the app being closed? Should all the Hosted Services get resumed when the app is resumed? These are all scenarios that would need to work and the Hosted Service abstraction isn't designed for these scenarios. Maui apps that need background service functionality need to use something that is designed for client apps.

5. We aren't using the Host that is built today. Start isn't even called on the object. And neither do we call Stop on it to shut it down properly if we did. Again this gets into the same lifetime questions as the Hosted Services. Should the Host be "Stopped" during deactivation? Extensions.Hosting isn't designed to be re-startable. Once it is stopped, it disposes of its services and you need to start over with the Builder to create a new one.

6. Lastly, Extensions.Hosting's API `IHostBuilder` has a "deferred execution" model. When `ConfigureAppConfiguration` or `ConfigureServices` is called, the code passed in doesn't run in-line. Instead the delegate passed is saved and only executed later during the `Build()` process. This isn't the easiest model for developers to learn and reason about how their code executes. And you can see us going away from these kind of API models with "[ASP.NET Minimal APIs](https://docs.microsoft.com/aspnet/core/fundamentals/minimal-apis)" and Maui is following the same pattern. When `MauiProgram.CreateMauiApp()` is executed, the code is executed in a more traditional "top down" approach. When you add a Service to the builder, the service is added now. After that line executes, the service is in the collection. The same for adding Configuration. There isn't this "two-phase configuration" step that we have in Extensions.Hosting. When you add a config source to the builder, you can immediately read config values from it. However, since MauiAppBuilder is trying to interact with the HostBuilder, it has a lot of workaround code to make sure this "deferred execution" model works correctly with the inline approach Maui. In case you are interested in the internals, check out [here](https://github.com/dotnet/maui/blob/27380ddcde911c4ca4729986a7581981dd8b9c04/src/Core/src/Hosting/Internal/BootstrapHostBuilder.cs#L13-L14), [here](https://github.com/dotnet/maui/blob/27380ddcde911c4ca4729986a7581981dd8b9c04/src/Core/src/Hosting/Internal/TrackingChainedConfigurationSource.cs#L5), and [here](https://github.com/dotnet/maui/blob/27380ddcde911c4ca4729986a7581981dd8b9c04/src/Core/src/Hosting/Internal/ConfigurationProviderSource.cs#L9).

With all this in mind, I propose we remove the dependency on `Microsoft.Extensions.Hosting` from Maui.

### Public API Changes

```diff
namespace Microsoft.Maui.Hosting
{
	public sealed class MauiAppBuilder
	{
-		public IHostBuilder Host { get; }
	}

-	public sealed class MauiApp : IHost
+	public sealed class MauiApp
	{
-		public IHostEnvironment Environment { get; }
-		public IHostApplicationLifetime Lifetime { get; }

-		public Task StartAsync(CancellationToken cancellationToken = default)
-		public Task StopAsync(CancellationToken cancellationToken = default)
-               void IDisposable.Dispose()
	}
}
```

The one scenario that still needs to be supported is replacing the DI container implementation with another implementation (for example Autofac, Unity, etc.). In order to keep supporting this, I propose we add the same method that `WebAssemblyHostBuilder` uses to enable this scenario:

```diff
namespace Microsoft.Maui.Hosting
{
	public sealed class MauiAppBuilder
	{
+		public void ConfigureContainer<TBuilder>(IServiceProviderFactory<TBuilder> factory, Action<TBuilder>? configure = null) where TBuilder : notnull
	}
}
```

### Intended Use-Case

As part of reducing startup cost and size of Maui applications, removing dependencies and code that aren't necessary is an easy way to improve startup and size.

I prototyped removing Extensions.Hosting, and [disabling Logging by default](https://github.com/dotnet/maui/issues/4394). These two combined changes resulted in a 13% reduction in time to start a `dotnet new maui` app on my Android emulator.

**Comment by eerhardt** [comment]
cc @Redth @Eilon @mattleibow @PureWeen - thoughts?

**Comment by Eilon** [comment]
I'm curious - does this change work with MAUI Blazor? A lot of the changes that were introduced here were to enable those scenarios and I want to make sure that whatever we do that we don't regress that.

Also, some of the items noted in this issue are not yet implemented, such as [hosted services](https://github.com/dotnet/maui/issues/2244). I think hosted services definitely make sense for client apps - it's just a way to start/stop things at the same time as the app, regardless of the app type. The issues related to it have several upvotes.

I think it would be a shame to remove this. I would personally prefer finding a way to make these things cheaper, or more of an opt-in if needed. At one point we tried copying and trimming the original code but it's _a lot_ of code so we ended up going with mostly re-using the original types. Perhaps it _is_ worth copying the code and trimming, and then in .NET 7 we re-work the original types in MS.Ext.Hosting to be more mobile-friendly.

Regarding measurements - I'm very glad we're bringing actual numbers to this. What proportion of the 13% is due to logging vs. the other changes? I think disabling logging in release is not very contentious - especially because any app can still do whatever it wants. But this proposal here is much more significant.

**Comment by mattleibow** [comment]
With regards to hosting and even just hosted services, how does WASM app handle this? If it is not used, then why would we need it then? 

I do think it is an important comparison, a Blazor WASM app is like a web app that is packaged up and sent to the client - just like a Maui Blazor app is.

If hosting is not used on WASM, was it a conscious decision to remove support for some reason, or is it literally not needed at all? As in, all the features we think we need from hosting is not actually used but rather something else is.

**Comment by davidbuckleyni** [comment]
Dont forget this will be shared with blazor and in future who knows if asp.net comes into the mix.

**Comment by Redth** [comment]
> 2. Extensions.Hosting has an "Environment" concept, which are things like "Development", "Staging", and "Production". And while this concept could map to client apps, the way these environments change in Hosted Apps is through Configuration - ex. environment variable, command line parameter, appsettings.json file, etc. My understanding is that a Mobile app doesn't get a chance to change these kinds of configurations once it is built. How do you change a .json file in your already built .apk file?

While you're right that the settings effectively happen at build time, I think there's still value in providing some implementation of `appsettings.json` as it's a familiar pattern and is useful for debug vs release environments.  It's sort of typical to use an `[EmbeddedResource]` of a json file for this purpose.. [I've done it](https://redth.codes/settings-json-files-in-xamarin-apps/) and [so has James](https://montemagno.com/add-asp-net-cores-dependency-injection-into-xamarin-apps-with-hostbuilder/#appsettingsjson) we even have a [community PR](https://github.com/dotnet/maui/pull/3872) on this.

The problem is, this isn't a super efficient thing to do, load an embedded resource, and parse the json at startup for values which are effectively static.

I keep thinking that some sort of source generator or build task could bridge the performance gap here, but that's not necessarily a small amount of effort to accomplish.  Is there something we could add in the short term that's less efficient of an implementation up front, but can be replaced with something source generator or build task based without changing behaviour or API between RC and GA?

Alternatively it is something that can be deferred until .NET 7 and left out for now.  Adding later is always easier than removing or fixing a half thought through implementation.

**Comment by Redth** [comment]
> With all this in mind, I propose we remove the dependency on Microsoft.Extensions.Hosting from Maui.

I think this is ultimately the correct choice, at the very least for .NET 6.  The other benefit is hopefully a decreased dependency graph as well.  Did you measure app size differences (especially on iOS) with these changes? I'd be curious if there were significant savings/decreased assemblies needed.

**Comment by eerhardt** [comment]
> does this change work with MAUI Blazor? 

I don't believe this change will affect MAUI Blazor at all. I've tested the following apps with my [local prototype change](https://github.com/eerhardt/maui/tree/RemoveHosting), and they all appear to work the same with my changes as they do without:

* https://github.com/dotnet/maui/tree/main/src/Controls/samples/Controls.Sample.SingleProject
* https://github.com/microsoft/dotnet-podcasts/tree/main/src/Mobile
* https://github.com/microsoft/dotnet-podcasts/tree/main/src/MobileBlazor

> Also, some of the items noted in this issue are not yet implemented, such as hosted services. I think hosted services definitely make sense for client apps - it's just a way to start/stop things at the same time as the app, regardless of the app type. The issues related to it have several upvotes.

I agree that a similar high-level scenario is valuable to client apps. I just don't believe that the Extensions.Hosting library provides the necessary APIs to enable this scenario correctly for client apps. "it's just a way to start/stop things at the same time as the app" doesn't take into account the different lifecycle stages of a mobile app, nor how, when, and where this logic runs.

When we enable a "background services" feature for Maui, I think we need a full requirements and design of all the scenarios that need to be supported. For example, looking through https://shinylib.net/jobs/getting-started#schedule, it has much more functionality than just "Start" and "Stop".

> At one point we tried copying and trimming the original code but it's a lot of code so we ended up going with mostly re-using the original types. Perhaps it is worth copying the code and trimming, and then in .NET 7 we re-work the original types in MS.Ext.Hosting to be more mobile-friendly.

In point (6) above I called out that in trying to fit the deferred execution programming model of Hosting into the in-line model that MauiProgram wants, we ended up copying a lot of code from ASP.NET's WebApplicationBuilder. ASP.NET needed to do this because they wanted to support backwards compatibility with apps built with earlier versions. Maui doesn't need to support earlier versions that were using the deferred execution programming model because there weren't any existing Maui apps doing this.

This workaround code [here](https://github.com/dotnet/maui/blob/27380ddcde911c4ca4729986a7581981dd8b9c04/src/Core/src/Hosting/Internal/BootstrapHostBuilder.cs#L13-L14), [here](https://github.com/dotnet/maui/blob/27380ddcde911c4ca4729986a7581981dd8b9c04/src/Core/src/Hosting/Internal/TrackingChainedConfigurationSource.cs#L5), and [here](https://github.com/dotnet/maui/blob/27380ddcde911c4ca4729986a7581981dd8b9c04/src/Core/src/Hosting/Internal/ConfigurationProviderSource.cs#L9) is non-trivial and contributes to the slow start up. Since this code is unnecessary, removing it is the easiest way to make it faster.

> What proportion of the 13% is due to logging vs. the other changes?

I'll measure each proposed change separately and report numbers back here.

**Comment by eerhardt** [comment]
> With regards to hosting and even just hosted services, how does WASM app handle this?
> If hosting is not used on WASM, was it a conscious decision to remove support for some reason, or is it literally not needed at all?

WASM apps don't use Extensions.Hosting. I'm not the best to answer "why" they don't, but I believe @SteveSandersonMS should be able to.

**Comment by SteveSandersonMS** [comment]
> WASM apps don't use Extensions.Hosting. I'm not the best to answer "why" they don't, but I believe @SteveSandersonMS should be able to.

Two main reasons come to mind:

 * Blazor WebAssembly has always needed to provide the most lean and streamlined runtime environment that it can. If we didn't strictly need to use a specific bit of infrastructure, we'd prefer not to.
 * Blazor WebAssembly runs in an environment that imposes different constraints than you'd get on server/desktop. For example, a server/desktop app can synchronously look at the filesystem and load a `.json` config file if it's there. Whereas for Blazor WebAssembly any such I/O would become an async network call and potential delay the startup process, so it has its own patterns that combine or parallelize such steps.

More generally, the more high-level a feature area is, the less universal it can be. Hosting abstractions are pretty high-level (compared with, say, JSON serialization) and it's not clear that it's desirable or wise to insist on a single hosting abstraction for all possible app types and runtime environments. I can't think of any cases where Blazor WebAssembly developers have expressed that it's a problem that it doesn't use Hosting.Extensions.

**Comment by eerhardt** [comment]
> What proportion of the 13% is due to logging vs. the other changes?

I spent some time getting numbers again today. Here is what I found using a `-c Release` build of a `dotnet new maui` app on my local x86_64 Pixel 5 emulator and using [this script](https://github.com/jonathanpeppers/maui-profiling/blob/main/scripts/profile.ps1) to get startup times. Each app was launched 10 times, and then the startup times were averaged. I ran the script twice for each app.

* Current `main`: 2.432s and  2.416s
* With Hosting, but [removed Logging](https://github.com/dotnet/maui/blob/678ed3f72ecf91faf2e55af7a588044f98b8512b/src/Core/src/Hosting/MauiAppBuilder.cs#L231-L269): 2.385s and 2.370s
* Removed [Hosting and Logging](https://github.com/eerhardt/maui/tree/RemoveHosting): 2.219s and 2.207s

So you can see removing Hosting should provide the bulk of the startup performance improvements. I'm also seeing a 9% increase now, instead of the 13% I was seeing before.

> Did you measure app size differences (especially on iOS) with these changes?

@Redth (or others), I'm not sure how to do this. Can you point me to some instructions on how to measure iOS size?  I tried publishing a `dotnet new maui` app: `dotnet publish -f net6.0-ios -c Release`, but that gave errors `error : A runtime identifier must be specified in order to publish this project.`. Then when I tried specifying `-r ios-arm64`, it said `error : The RuntimeIdentifier 'ios-arm64' is invalid.`. I also don't know how to connect to a mac from the command line. Should I be using VS to measure the iOS size?

**Comment by eerhardt** [comment]
> Did you measure app size differences (especially on iOS) with these changes?

I finally got it working on my mac today. Here is what I measured:

1. `dotnet new maui`
2. `dotnet publish -c Release -f net6.0-ios -r ios-arm64`
3. Inspect `bin\Release\net6.0-ios\ios-arm64\publish\*.ipa` file size

`main` as of last Friday: `19.2 MB (20,176,526 bytes)`
Same commit with my [proposed change](https://github.com/eerhardt/maui/tree/RemoveHosting): `18.0 MB (18,903,808 bytes)`

So it appears to shave off ~1.2 MBs for the `dotnet new` app.
